src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/ImplementGenericMathInterfacesCorrectly.cs

```diff
@@ -0,0 +1,136 @@
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(GMIRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIParsable1, out var iParsableInterface) ||
+                    !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemNumericsINumber1, out var iNumberInterface))
+                {
+                    return;
+                }
+
+                context.RegisterSymbolAction(context =>
+                {
+                    if (context.Symbol is INamedTypeSymbol ntSymbol && !ntSymbol.IsGenericType)
+                    {
+                        AnalyzeSymbol(context, ntSymbol, iParsableInterface.ContainingNamespace, iNumberInterface.ContainingNamespace);
+                    }
+                }, SymbolKind.NamedType);
+            });
+        }
+
+        private void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol symbol, INamespaceSymbol systemNS, INamespaceSymbol systemNumericsNS)
+        {
+            foreach (INamedTypeSymbol anInterface in symbol.Interfaces)
+            {
+                if (IsKnownInterfaceInTheChain(anInterface) &&
+                    FirstTypeParameterNameIsNotTheSymbolName(symbol, anInterface))
+                {
+                    SyntaxNode? typeParameter = FindTheTypeArgumentOfTheInterfaceFromTypeDeclaration(symbol, anInterface);
+                    context.ReportDiagnostic(CreateDiagnostic(GMIRule, typeParameter, anInterface.OriginalDefinition.ToDisplayString(
+                        SymbolDisplayFormat.MinimallyQualifiedFormat), anInterface.OriginalDefinition.TypeParameters[0].Name, symbol));
+                    break;
+                }
+            }
+
+            INamedTypeSymbol? baseType = symbol.BaseType;
+            while (baseType != null && baseType.IsGenericType)
+            {
+                foreach (INamedTypeSymbol anInterface in baseType.Interfaces)
+                {
+                    if (IsKnownInterfaceInTheChain(anInterface) &&
+                        FirstTypeParameterNameIsNotTheSymbolName(symbol, anInterface))
+                    {
+                        SyntaxNode? typeParameter = FindTheTypeArgumentOfTheInterfaceFromTypeDeclaration(symbol, symbol.BaseType);
+                        context.ReportDiagnostic(CreateDiagnostic(GMIRule, typeParameter, symbol.BaseType.OriginalDefinition.ToDisplayString(
+                            SymbolDisplayFormat.MinimallyQualifiedFormat), symbol.BaseType.OriginalDefinition.TypeParameters[0].Name, symbol));
+                        break;
+                    }
+                }
```

Comment:
Are usages of `symbol.BaseType` here correct? It looks like we're walking the hierarchy with `baseType = baseType.BaseType;`, so should these be `baseType` instead of `symbol.BaseType`?

Reply:
Its correct, we want to show the diagnostic on the `symbol.BaseType` only, iterating by `baseType.BaseType` only to check if any of `symbol.BaseType`'s  parents have implemented a GM interface